Runtime/Tiles/RandomTerrainTile/RandomTerrainTile.cs

```diff
@@ -0,0 +1,349 @@
+                for (int xd = -1; xd <= 1; xd++)
+                {
+                    Vector3Int position = new Vector3Int(location.x + xd, location.y + yd, location.z);
+                    if (TileValue(tileMap, position))
+                        tileMap.RefreshTile(position);
+                }
+        }
+
+        /// <summary>
+        /// Retrieves any tile rendering data from the scripted tile.
+        /// </summary>
+        /// <param name="position">Position of the Tile on the Tilemap.</param>
+        /// <param name="tilemap">The Tilemap the tile is present on.</param>
+        /// <param name="tileData">Data to render the tile.</param>
+        public override void GetTileData(Vector3Int location, ITilemap tileMap, ref TileData tileData)
+        {
+            UpdateTile(location, tileMap, ref tileData);
+        }
+
+        private void UpdateTile(Vector3Int location, ITilemap tileMap, ref TileData tileData)
+        {
+            tileData.transform = Matrix4x4.identity;
+            tileData.color = Color.white;
+
+            int mask = TileValue(tileMap, location + new Vector3Int(0, 1, 0)) ? 1 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(1, 1, 0)) ? 2 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(1, 0, 0)) ? 4 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(1, -1, 0)) ? 8 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(0, -1, 0)) ? 16 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(-1, -1, 0)) ? 32 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(-1, 0, 0)) ? 64 : 0;
+            mask += TileValue(tileMap, location + new Vector3Int(-1, 1, 0)) ? 128 : 0;
+
+            byte original = (byte)mask;
+            if ((original | 254) < 255) { mask = mask & 125; }
+            if ((original | 251) < 255) { mask = mask & 245; }
+            if ((original | 239) < 255) { mask = mask & 215; }
+            if ((original | 191) < 255) { mask = mask & 95; }
+
+            Sprite[][] m_Sprites = GetSpritesAsOne();
+
+            int index = GetIndex((byte)mask);
+            if (index >= 0 && index < m_Sprites.Length && TileValue(tileMap, location))
+            {
+                tileData.sprite = GetRandomSprite(m_Sprites[index]);
+                tileData.transform = GetTransform((byte)mask);
+                tileData.color = Color.white;
+                tileData.flags = TileFlags.LockTransform | TileFlags.LockColor;
+                tileData.colliderType = Tile.ColliderType.Sprite;
+            }
+        }
+
+        private Sprite GetRandomSprite(Sprite[] spriteList) {
+            if (spriteList.Length == 0) {
+                return null;
+            }
+
+            int spriteIndex = Random.Range(0, spriteList.Length - 1);
+            return spriteList[spriteIndex];
+        }
+
+        private Sprite[][] GetSpritesAsOne() {
+            Sprite[][] m_Sprites = new Sprite[15][];
```

Comment:
It looks like this could be a switch statement returning the correct Sprite Array when the index is passed it. This would avoid creating a new array of arrays each time.